### PR TITLE
GEOMESA-1334 OSM XML/PBF converters

### PR DESCRIPTION
**Pull request by elahrvivaz**
- Supports 'nodes' and 'ways' but not 'relations'
- All ways are treated as linestrings
- References to missing nodes are dropped, but the way is still created

Signed-off-by: Emilio Lahr-Vivaz elahrvivaz@ccri.com

### CODE REVIEW

**Comment by jahhulbert-ccri** [comment]
we still waiting on a CQ?

File: build/cqs.tsv [hunk 1]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
...
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
+de.topobyte:jsi	1.1.0	compile
+de.topobyte:jts-utils	0.0.4	compile
+de.topobyte:large-scale-file-io	0.0.1	compile
+de.topobyte:melon-io	0.0.1	compile
+de.topobyte:osm4j-core	0.0.18	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11792

File: build/cqs.tsv [hunk 2]
```diff
@@ -15,6 +15,8 @@ com.googlecode.juniversalchardet:juniversalchardet	1.0.3	compile
 com.google.guava:guava	14.0.1	compile
 com.google.guava:guava	17.0	compile
 com.google.guava:guava	18.0	compile
+com.google.protobuf:protobuf-java	2.6.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11790

File: build/cqs.tsv [hunk 3]
```diff
@@ -15,6 +15,8 @@ com.googlecode.juniversalchardet:juniversalchardet	1.0.3	compile
...
+com.h2database:h2	1.4.192	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11791

File: build/cqs.tsv [hunk 4]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
...
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
+de.topobyte:jsi	1.1.0	compile
+de.topobyte:jts-utils	0.0.4	compile
+de.topobyte:large-scale-file-io	0.0.1	compile
+de.topobyte:melon-io	0.0.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11793

File: build/cqs.tsv [hunk 5]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
...
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
+de.topobyte:jsi	1.1.0	compile
+de.topobyte:jts-utils	0.0.4	compile
+de.topobyte:large-scale-file-io	0.0.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11794

File: build/cqs.tsv [hunk 6]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
...
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
+de.topobyte:jsi	1.1.0	compile
+de.topobyte:jts-utils	0.0.4	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11795

File: build/cqs.tsv [hunk 7]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
 com.typesafe.scala-logging:scala-logging_2.11	3.1.0	compile
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
+de.topobyte:jsi	1.1.0	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11796

File: build/cqs.tsv [hunk 8]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
 com.typesafe.scala-logging:scala-logging_2.11	3.1.0	compile
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11797

File: build/cqs.tsv [hunk 9]
```diff
@@ -36,6 +38,22 @@ com.typesafe:config	1.2.1	compile
 com.typesafe.scala-logging:scala-logging_2.11	3.1.0	compile
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:commons-cli-helper	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11798

File: build/cqs.tsv [hunk 10]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
 com.typesafe.scala-logging:scala-logging_2.11	3.1.0	compile
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
+de.topobyte:dynsax	0.0.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11797

File: build/cqs.tsv [hunk 11]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
 com.typesafe.scala-logging:scala-logging_2.11	3.1.0	compile
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
+de.topobyte:adt-graph	0.0.1	compile
+de.topobyte:adt-multicollections	0.0.3	compile
+de.topobyte:compact-io	0.0.5	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11798

File: build/cqs.tsv [hunk 12]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
...
+de.topobyte:dynsax	0.0.1	compile
+de.topobyte:jsi	1.1.0	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11796

File: build/cqs.tsv [hunk 13]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
...
+de.topobyte:jts-utils	0.0.4	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11795

File: build/cqs.tsv [hunk 14]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
 com.typesafe.scala-logging:scala-logging_2.11	3.1.0	compile
 com.vividsolutions:jts	1.13	compile
 com.yammer.metrics:metrics-core	2.2.0	compile
+de.topobyte:adt-geo	0.0.2	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11799

File: build/cqs.tsv [hunk 15]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
...
+de.topobyte:adt-graph	0.0.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11800

File: build/cqs.tsv [hunk 16]
```diff
@@ -36,6 +38,21 @@ com.typesafe:config	1.2.1	compile
...
+de.topobyte:adt-multicollections	0.0.3	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11801

File: build/cqs.tsv [hunk 17]
```diff
@@ -70,16 +87,17 @@ it.geosolutions.jaiext.utilities:jt-utilities	1.0.8	compile
 it.geosolutions.jaiext.vectorbin:jt-vectorbin	1.0.8	compile
 it.geosolutions.jaiext.warp:jt-warp	1.0.8	compile
 it.geosolutions.jaiext.zonal:jt-zonal	1.0.8	compile
-it.unimi.dsi:fastutil	7.0.6	compile
 joda-time:joda-time	2.3	compile
 log4j:log4j	1.2.17	compile
 net.iharder:base64	2.3.8	compile
 net.java.dev.jsr-275:jsr-275	1.0-beta-2	compile
 net.jpountz.lz4:lz4	1.2.0	compile
+net.jpountz.lz4:lz4	1.3.0	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11802

File: build/cqs.tsv [hunk 18]
```diff
@@ -70,16 +87,17 @@ it.geosolutions.jaiext.utilities:jt-utilities	1.0.8	compile
...
 net.minidev:asm	1.0.2	compile
 net.minidev:json-smart	2.1.1	compile
 net.sf.geographiclib:GeographicLib-Java	1.44	compile
 net.sf.jopt-simple:jopt-simple	3.2	compile
+net.sf.trove4j:trove4j	3.0.3	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11803

**Comment by elahrvivaz** [comment]
cqs added

**Comment by elahrvivaz** [comment]
i'm testing this out by ingesting the `ways` out of the whole `planet-latest.osm.pbf`, but the h2 database is really starting to struggle with ingesting the nodes. it's up to 72G of files, but it's slowed down to a crawl... if no one has any bright ideas, I'll open a ticket to address it later.

**Comment by jahhulbert-ccri** [comment]
how big is the h2 db? i wonder if a hash index as the primary key would be better...depends on how much mem you have but would give direct lookups.

**Comment by elahrvivaz** [comment]
i was trying not to rely on main mem - the h2 got to be 70G+ on disk... i think i might have figured out the problem though - i was using the node id as the primary key, but h2 says that it keeps files sorted by primary key, so any insert would likely touch half the files. I'm trying again without any primary key (so it will use a 1-up internal id), then creating an index on the id field after all the nodes are ingested.

**Comment by jahhulbert-ccri** [comment]
gotcha yeah i think most do a clustered database on the primary key

**Comment by elahrvivaz** [comment]
going to try a few other things - want to be able to ingest the whole planet-osm file.

File: geomesa-convert/geomesa-convert-osm/src/main/scala/org/locationtech/geomesa/convert/osm/OsmWaysConverter.scala [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+                       val needsMetadata: Boolean) extends ToSimpleFeatureConverter[OsmWay] with LazyLogging {
+
+  private def gf = JTSFactoryFinder.getGeometryFactory
+  private val toArray = if (needsMetadata) { OsmField.toArrayWithMetadata _ } else { OsmField.toArrayNoMetadata _ }
+
+  // create a temporary h2 database to store our nodes
+  // the ways only refer to nodes by reference, but we can't keep the whole file in memory
+  private val h2Dir = Files.createTempDirectory("geomesa-convert-h2").toFile
+  private val connection = {
+    Class.forName("org.h2.Driver")
```
Comment (jnh5y) [inline geomesa-convert/geomesa-convert-osm/src/main/scala/org/locationtech/geomesa/convert/osm/OsmWaysConverter.scala]: Is this line needed:)?

File: geomesa-convert/geomesa-convert-osm/src/test/scala/org/locationtech/geomesa/convert/osm/OsmConverterLoadTest.scala [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+        """
+          | {
+          |   type         = "osm-ways"
+          |   id-field     = "$id"
+          |   fields = [
+          |     { name = "id", attribute = "id", transform = "toString($0)" }
+          |     { name = "user", attribute = "user" }
+          |     { name = "tags", attribute = "tags" }
+          |     { name = "name", transform = "mapValue($tags, 'name')" }
+          | //  { name = "name", attribute = "tags" transform = "mapValue($0, 'name')" }
```
Comment (jnh5y) [inline geomesa-convert/geomesa-convert-osm/src/test/scala/org/locationtech/geomesa/convert/osm/OsmConverterLoadTest.scala]: Is this commented line necessary?

File: geomesa-convert/geomesa-convert-osm/src/main/scala/org/locationtech/geomesa/convert/osm/OsmWaysConverter.scala (hunk 19 above)
Comment (elahrvivaz) [inline geomesa-convert/geomesa-convert-osm/src/main/scala/org/locationtech/geomesa/convert/osm/OsmWaysConverter.scala]: yeah, this is the typical pattern used for jdbc to ensure that the classpath is correct I guess

File: geomesa-convert/geomesa-convert-osm/src/test/scala/org/locationtech/geomesa/convert/osm/OsmConverterLoadTest.scala (hunk 20 above)
Comment (elahrvivaz) [inline geomesa-convert/geomesa-convert-osm/src/test/scala/org/locationtech/geomesa/convert/osm/OsmConverterLoadTest.scala]: mmm, probably not

**Comment by jnh5y** [comment]
@elahrvivaz, apologies for the delays in getting the CQs approved.  In case that continues, are there bits of this PR which should be merged (like the DeleteFileVisitor)?

If there are issues with the CQs, would the work be able to be committed to a module/tutorial repo?

**Comment by elahrvivaz** [comment]
I need to go back and fix this - all the embedded databases I've tried can't handle the whole-planet osm file for the `ways` ingestion. I've run out of ideas at this point... I don't think there are any parts worth separating out. moving to an alternate repo might be an idea but waiting for cqs is ok for now.

**Comment by jnh5y** [comment]
@elahrvivaz yeah... I'm suggesting the alternate repo since I think getting the CQs approved might take quite a bit of work...

Is there a sensible way to use Accumulo for managing the intermediate database?  All the lookups for nodes should be fast...

**Comment by elahrvivaz** [comment]
yeah, using a 'real' db would probably fix the problem. i was hesitant to require something like that in the converter framework though. might be the way to go...

**Comment by jahhulbert-ccri** [comment]
@elahrvivaz what if we allowed a jdbc string for a driver instead? so if you wanted to do it you could?

**Comment by elahrvivaz** [comment]
@jahhulbert-ccri yeah, good idea

File: build/cqs.tsv (hunk 1 above)
Comment (jahhulbert-ccri) [inline build/cqs.tsv]: i think its ok now

File: build/cqs.tsv (hunk 4 above)
Comment (jahhulbert-ccri) [inline build/cqs.tsv]: looks approved now

**Comment by jahhulbert-ccri** [comment]
ah looks like it needs a little master merge too.

**Comment by elahrvivaz** [comment]
looks like one more cq: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11802

**Comment by jahhulbert-ccri** [comment]
What! it's done?!?!!? one more build/cqs possibly then we can merge.

**Comment by jnh5y** [comment]
Yes.  There is light at the end of the tunnel!  Looks like we need to re-run the build-cqs.tsv, resolve conflicts and shippit!

**Comment by jahhulbert-ccri** [comment]
yay! cool once @elahrvivaz get's that in I'll merge. There's some other OSM GPX code I need to get in that apparently is in my repo and I'm going to pull it over: https://geomesa.atlassian.net/browse/GEOMESA-1730

**Comment by elahrvivaz** [comment]
rebased and double-checked cqs, think we should be good now

**Pull request merged**